src/theory/quantifiers/sygus/example_infer.cpp

```diff
@@ -0,0 +1,266 @@
+ **
+ ** \brief Implementation of utility for inferring whether a formula is in
+ ** examples form (functions applied to concrete arguments only).
+ **
+ **/
+#include "theory/quantifiers/sygus/example_infer.h"
+
+#include "theory/quantifiers/quant_util.h"
+
+using namespace CVC4;
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+ExampleInfer::ExampleInfer(TermDbSygus* tds) : d_tds(tds)
+{
+  d_isExamples = false;
+}
+
+ExampleInfer::~ExampleInfer() {}
+
+bool ExampleInfer::initialize(Node n, const std::vector<Node>& candidates)
+{
+  Trace("ex-infer") << "Initialize example inference : " << n << std::endl;
+
+  for (const Node& v : candidates)
+  {
+    d_examples[v].clear();
+    d_examplesOut[v].clear();
+    d_examplesTerm[v].clear();
+  }
+
+  std::map<Node, bool> visited;
+  // n is negated conjecture
+  if (!collectExamples(n, visited, true, false))
+  {
+    Trace("ex-infer") << "...conflicting examples" << std::endl;
+    return false;
+  }
+
+  if (Trace.isOn("ex-infer"))
+  {
+    for (unsigned i = 0; i < candidates.size(); i++)
+    {
+      Node v = candidates[i];
+      Trace("ex-infer") << "  examples for " << v << " : ";
+      if (d_examples_invalid.find(v) != d_examples_invalid.end())
+      {
+        Trace("ex-infer") << "INVALID" << std::endl;
+      }
+      else
+      {
+        Trace("ex-infer") << std::endl;
+        for (unsigned j = 0; j < d_examples[v].size(); j++)
+        {
+          Trace("ex-infer") << "    ";
+          for (unsigned k = 0; k < d_examples[v][j].size(); k++)
+          {
+            Trace("ex-infer") << d_examples[v][j][k] << " ";
+          }
+          if (!d_examplesOut[v][j].isNull())
+          {
+            Trace("ex-infer") << " -> " << d_examplesOut[v][j];
+          }
+          Trace("ex-infer") << std::endl;
+        }
+      }
+      Trace("ex-infer") << "Initialize " << d_examples[v].size()
+                        << " example points for " << v << "..." << std::endl;
+    }
+  }
+  return true;
+}
+
+bool ExampleInfer::collectExamples(Node n,
+                                   std::map<Node, bool>& visited,
+                                   bool hasPol,
+                                   bool pol)
+{
+  if (visited.find(n) != visited.end())
+  {
+    // already visited
+    return true;
+  }
+  visited[n] = true;
+  NodeManager* nm = NodeManager::currentNM();
+  Node neval;
+  Node n_output;
+  bool neval_is_evalapp = false;
+  if (n.getKind() == DT_SYGUS_EVAL)
+  {
+    neval = n;
+    if (hasPol)
+    {
+      n_output = nm->mkConst(pol);
+    }
+    neval_is_evalapp = true;
+  }
+  else if (n.getKind() == EQUAL && hasPol && pol)
+  {
+    for (unsigned r = 0; r < 2; r++)
+    {
+      if (n[r].getKind() == DT_SYGUS_EVAL)
+      {
+        neval = n[r];
+        if (n[1 - r].isConst())
+        {
+          n_output = n[1 - r];
+        }
+        neval_is_evalapp = true;
+      }
+    }
```

Comment:
This does not seem robust for `f(c1) = f(c2)`. How is that being handled?

Reply:
You're right, there is a bug (in the existing code as well)
Will fix.

Reply:
If we are not equality but not an I/O pair (one side constant), we now always traverse to the subterms, this means we will collect both c1 and c2.

Reply:
I don't get it. :)

Reply:
This code sets neval_is_evalapp to true if we are f( s ) = t for any s, t.  It may be that s, t are not constants.  It doesnt matter if t is a non-constant (x) or another application of a function to synthesize, the check for isConst on line 118 here will fail.

Reply:
I see. Another thing: in the case of `f(c1) = f(c2)` isn't the code gonna add the example `c2` twice? Once in the code block starting at line 127 and then when it visits the child `f(c2)` from the recursive call at line 193.

Reply:
I've updated the code so that this is explicitly guarded against.  The previous code was indeed unclear whether this could happen.

Reply:
Cool, thanks!